### PR TITLE
feat: added take-along labels handling

### LINKED ISSUE #221: Take along cluster resource labels
Hi!

it would be nice if certain labels of the cluster resource can be taken along to the ArgoCD secret. This can be useful if you've configured ArgoCD to deploy certain stuff only on clusters with certain labels, e.g. `env: stage` or `env: prod`.

Could be configurable via a label in the cluster resource like `take-along-label.capi-to-argocd.<key>=`
One has to read the cluster resource and get the labels and put them on the ArgoCD secret according to the take-along label.

Example for the cluster resource: 
```yaml
apiVersion: cluster.x-k8s.io/v1beta1
kind: Cluster
metadata: 
  labels: 
    cloudprovider.clusters.infrastructure.edgefarm.io/type: hetzner
    clusters.infrastructure.edgefarm.io/type: core
    core.edgefarm.io/name: dog
    env: stage
    deploy: group1
    foo: bar
    take-along-label.capi-to-argocd.env: ""
    take-along-label.capi-to-argocd.deploy: ""
  name: dog
  namespace: default
...
```

And the resulting secret:

```yaml
kind: Secret
metadata: 
  labels: 
    argocd.argoproj.io/secret-type: cluster
    capi-to-argocd/cluster-namespace: coreclusters
    capi-to-argocd/cluster-secret-name: dog-kubeconfig
    capi-to-argocd/owned: "true"
    env: stage
    deploy: group1
  name: cluster-dog
  namespace: argocd
...
```

@dntosas what do you think?

Update: changed `take-along-label.capi-to-argocd/<key>` to `take-along-label.capi-to-argocd-<key>`, because k8s label prevents from setting multiple `/`. This is the case when taking a long label along e.g. `my.domain.com/subdomain`. Using the `/` as delimiter it would look like this `take-along-label.capi-to-argocd/my.domain.com/subdomain` which is invalid.

### PR DESCRIPTION
**Pull request by siredmar**
fixes #221

[x] implementation
[x] unit tests
[x] docs

### CODE REVIEW

**Comment by siredmar** [comment]
Unit tests and docs will WIP. Feel free to leave a review.

**Comment by dntosas** [comment]
@siredmar thnx for the addition! will take a look in monday unless that's real urgent for you ^

nit: if you could split vendor files to a dedicated commit, could possible make review much easier

**Comment by siredmar** [comment]
@dntosas no problem with monday :) 
i''ve split the commits in vendor updates and the change itself.

**Comment by siredmar** [comment]
@dntosas i'm struggling with adding the reconcile test. This PR does not has working tests :/ 
I hope i can figure this out anytime soon. Maybe you can take a look too? I've created an issue that describes the problem https://github.com/kubernetes-sigs/controller-runtime/issues/2530

The implementation itself works in a cluster, the Reconceile unit test does not.

File: controllers/capi2argo_reconciler_test.go [hunk 1]
```diff
@@ -40,15 +53,27 @@ var _ = BeforeSuite(func() {
 	By("bootstrapping test environment")
 	logf.SetLogger(zap.New(zap.WriteTo(GinkgoWriter), zap.UseDevMode(true)))
 
-	TestEnv = &envtest.Environment{}
+	Scheme = runtime.NewScheme()
+	gv := clusterv1.GroupVersion
+	Scheme.AddKnownTypes(gv, &clusterv1.Cluster{}, &clusterv1.ClusterList{})
+	utilruntime.Must(clientgoscheme.AddToScheme(Scheme))
+	utilruntime.Must(apiextensionsv1.AddToScheme(Scheme))
+	_ = clusterv1alpha4.AddToScheme(Scheme)
```
Comment (troy0820) [inline controllers/capi2argo_reconciler_test.go:61]:
```suggestion
	_ = clusterv1alpha4.AddToScheme(Scheme)
	utilruntime.Must(cluterv1.AddToScheme)
```
Your issue seems to be that you do not have the clutser's cluster-api scheme added.  

This should help get your tests passing.

**Review by troy0820 — CHANGES_REQUESTED** [review]
Add the scheme for cluster-api so you can convert the cluster from the test failure in controller-runtime issue https://github.com/kubernetes-sigs/controller-runtime/issues/2530

https://pkg.go.dev/sigs.k8s.io/cluster-api/api/v1beta1#Cluster

File: controllers/capi2argo_reconciler_test.go (hunk 1 above)
Comment (siredmar) [reply controllers/capi2argo_reconciler_test.go:61]: @troy0820 thx for taking a look! I already tried this with no luck. line 58 and above is another approach of registering the correct schemes.

**Comment by troy0820** [comment]
image
Doing a make test gave me the following output with the changes below: 
```go
 var _ = BeforeSuite(func() {
            By("bootstrapping test environment")
            logf.SetLogger(zap.New(zap.WriteTo(GinkgoWriter), zap.UseDevMode(true)))
   
            Scheme = runtime.NewScheme()
   
            TestEnv = &envtest.Environment{
                    CRDDirectoryPaths: []string{
                            filepath.Join("..", "tests", "vendor", "cluster-api", "crd", "v1beta"),
                    },
                    ErrorIfCRDPathMissing:   true,
                    ControlPlaneStopTimeout: time.Second * 100,
            }
            cfg, err := TestEnv.Start()
            Expect(err).NotTo(HaveOccurred())
            Expect(cfg).NotTo(BeNil())
   
            utilruntime.Must(clusterv1.AddToScheme(Scheme))
            utilruntime.Must(clientgoscheme.AddToScheme(Scheme))
```

I hope this helps but I'm not recieving the error you posted.

**Comment by troy0820** [comment]
> image Doing a make test gave me the same error with the following changes below:
> 
> ```go
>  var _ = BeforeSuite(func() {
>             By("bootstrapping test environment")
>             logf.SetLogger(zap.New(zap.WriteTo(GinkgoWriter), zap.UseDevMode(true)))
>    
>             Scheme = runtime.NewScheme()
>    
>             TestEnv = &envtest.Environment{
>                     CRDDirectoryPaths: []string{
>                             filepath.Join("..", "tests", "vendor", "cluster-api", "crd", "v1beta"),
>                     },
>                     ErrorIfCRDPathMissing:   true,
>                     ControlPlaneStopTimeout: time.Second * 100,
>             }
>             cfg, err := TestEnv.Start()
>             Expect(err).NotTo(HaveOccurred())
>             Expect(cfg).NotTo(BeNil())
>    
>             utilruntime.Must(clusterv1.AddToScheme(Scheme))
>             utilruntime.Must(clientgoscheme.AddToScheme(Scheme))
> ```
> 
> The error posted is on line 102 where you get the cluster object
> 
> ```go
>  err := MockReconcileEnv()
>  assert.Nil(t, err)
>  
>  cluster, err := getClusterObject("test", TestNamespace, Scheme)
> ```
> 
> When you do this, there is no cluster object to get and it will fail when asserting the type that doesn't exists.

In that function you are making 2 namespaces, 3 secrets and getting a cluster that doesn't exist yet (so it doesn't fail because isNotFound and then gets created)

After this happens, you immediately get the created cluster that seems to have an issue with what was just written.  I would make sure that you can get the cluster where the k8sClient is not pulling from cache if you need to do something with it immediately.

**Comment by troy0820** [comment]
```python
make test
KUBEBUILDER_ASSETS="/home/troy0820/workspace/capi2argo-cluster-operator/bin/k8s/1.26.1-linux-amd64" go test -v -mod=vendor `go list ./...` -coverprofile cover.out
?       github.com/dntosas/capi2argo-cluster-operator   [no test files]
=== RUN   TestExtractTakeAlongLabel
=== PAUSE TestExtractTakeAlongLabel
=== RUN   TestConvertToSecret
=== PAUSE TestConvertToSecret
=== RUN   TestValidateClusterTLSConfig
=== PAUSE TestValidateClusterTLSConfig
=== RUN   TestBuildNamespacedName
=== PAUSE TestBuildNamespacedName
=== RUN   TestControllers
Running Suite: Capi2ArgoClusterOperator Controller Suite
========================================================
Random Seed: 1696873150
Will run 0 of 0 specs


Ran 0 of 0 Specs in 63.740 seconds
SUCCESS! -- 0 Passed | 0 Failed | 0 Pending | 0 Skipped
--- PASS: TestControllers (63.74s)
=== RUN   TestReconcile
    capi2argo_reconciler_test.go:109:
                Error Trace:    /home/troy0820/workspace/capi2argo-cluster-operator/controllers/capi2argo_reconciler_test.go:109
                Error:          Expected nil, but got: &url.Error{Op:"Post", URL:"https://127.0.0.1:36509/api/v1/namespaces/test/secrets", Err:(*net.OpError)(0xc00051a320)}
                Test:           TestReconcile
=== RUN   TestReconcile/process_valid_secret
=== PAUSE TestReconcile/process_valid_secret
=== RUN   TestReconcile/process_existing_valid_secret
=== PAUSE TestReconcile/process_existing_valid_secret
=== RUN   TestReconcile/process_secret_with_wrong_Data[key]
=== PAUSE TestReconcile/process_secret_with_wrong_Data[key]
=== RUN   TestReconcile/process_secret_with_wrong_Type
=== PAUSE TestReconcile/process_secret_with_wrong_Type
=== CONT  TestReconcile/process_valid_secret
=== CONT  TestReconcile/process_secret_with_wrong_Data[key]
=== CONT  TestReconcile/process_existing_valid_secret
=== CONT  TestReconcile/process_secret_with_wrong_Type
=== NAME  TestReconcile/process_secret_with_wrong_Data[key]
    capi2argo_reconciler_test.go:155:
                Error Trace:    /home/troy0820/workspace/capi2argo-cluster-operator/controllers/capi2argo_reconciler_test.go:155
                Error:          Expected value not to be nil.
                Test:           TestReconcile/process_secret_with_wrong_Data[key]
    capi2argo_reconciler_test.go:156:
                Error Trace:    /home/troy0820/workspace/capi2argo-cluster-operator/controllers/capi2argo_reconciler_test.go:156
                Error:          An error is expected but got nil.
                Test:           TestReconcile/process_secret_with_wrong_Data[key]
=== NAME  TestReconcile/process_secret_with_wrong_Type
    capi2argo_reconciler_test.go:155:
                Error Trace:    /home/troy0820/workspace/capi2argo-cluster-operator/controllers/capi2argo_reconciler_test.go:155
                Error:          Expected value not to be nil.
                Test:           TestReconcile/process_secret_with_wrong_Type
    capi2argo_reconciler_test.go:156:
                Error Trace:    /home/troy0820/workspace/capi2argo-cluster-operator/controllers/capi2argo_reconciler_test.go:156
                Error:          An error is expected but got nil.
                Test:           TestReconcile/process_secret_with_wrong_Type
--- FAIL: TestReconcile (5.00s)
    --- PASS: TestReconcile/process_existing_valid_secret (0.00s)
    --- PASS: TestReconcile/process_valid_secret (0.00s)
    --- FAIL: TestReconcile/process_secret_with_wrong_Data[key] (0.00s)
    --- FAIL: TestReconcile/process_secret_with_wrong_Type (0.00s)
=== RUN   TestValidateObjectOwner
--- PASS: TestValidateObjectOwner (0.00s)
=== RUN   TestUnmarshal
=== PAUSE TestUnmarshal
=== RUN   TestNewCapiCluster
--- PASS: TestNewCapiCluster (0.00s)
=== RUN   TestValidateCapiSecret
=== PAUSE TestValidateCapiSecret
=== CONT  TestExtractTakeAlongLabel
=== RUN   TestExtractTakeAlongLabel/Test_with_valid_take-along-labels_label
=== PAUSE TestExtractTakeAlongLabel/Test_with_valid_take-along-labels_label
=== CONT  TestBuildNamespacedName
=== CONT  TestConvertToSecret
=== CONT  TestValidateCapiSecret
=== CONT  TestUnmarshal
=== RUN   TestConvertToSecret/test_type_with_valid_fields
=== RUN   TestExtractTakeAlongLabel/Test_with_complex_and_valid_take-along-labels_label
=== RUN   TestBuildNamespacedName/test_type_with_valid_fields
=== CONT  TestValidateClusterTLSConfig
=== RUN   TestUnmarshal/test_type_with_valid_fields
=== PAUSE TestUnmarshal/test_type_with_valid_fields
=== RUN   TestUnmarshal/test_type_with_wrong_secret.Data[key]
=== PAUSE TestUnmarshal/test_type_with_wrong_secret.Data[key]
=== RUN   TestUnmarshal/test_type_with_wrong_secret.Type
=== PAUSE TestUnmarshal/test_type_with_wrong_secret.Type
=== PAUSE TestExtractTakeAlongLabel/Test_with_complex_and_valid_take-along-labels_label
=== RUN   TestExtractTakeAlongLabel/Test_with_no_take-along-labels_labels
=== RUN   TestValidateClusterTLSConfig/test_type_with_valid_fields
=== PAUSE TestExtractTakeAlongLabel/Test_with_no_take-along-labels_labels
=== PAUSE TestValidateClusterTLSConfig/test_type_with_valid_fields
=== RUN   TestValidateClusterTLSConfig/test_type_with_non-valid_field
=== PAUSE TestValidateClusterTLSConfig/test_type_with_non-valid_field
=== RUN   TestValidateClusterTLSConfig/test_type_with_missing_fields
=== PAUSE TestValidateClusterTLSConfig/test_type_with_missing_fields
=== CONT  TestUnmarshal/test_type_with_valid_fields
=== RUN   TestValidateClusterTLSConfig/test_empty_type
=== PAUSE TestValidateClusterTLSConfig/test_empty_type
=== CONT  TestValidateClusterTLSConfig/test_type_with_valid_fields
=== RUN   TestExtractTakeAlongLabel/Test_with_standard_label
=== PAUSE TestExtractTakeAlongLabel/Test_with_standard_label
=== RUN   TestValidateCapiSecret/test_type_with_valid_fields
=== PAUSE TestConvertToSecret/test_type_with_valid_fields
=== CONT  TestUnmarshal/test_type_with_wrong_secret.Type
=== RUN   TestConvertToSecret/test_type_with_non-valid_fields
=== PAUSE TestConvertToSecret/test_type_with_non-valid_fields
=== CONT  TestValidateClusterTLSConfig/test_type_with_non-valid_field
=== RUN   TestBuildNamespacedName/test_type_with_valid_fields_and_namespaced_names
=== CONT  TestExtractTakeAlongLabel/Test_with_standard_label
=== CONT  TestExtractTakeAlongLabel/Test_with_no_take-along-labels_labels
=== RUN   TestBuildNamespacedName/test_type_with_non-valid_fields
=== CONT  TestExtractTakeAlongLabel/Test_with_complex_and_valid_take-along-labels_label
=== CONT  TestConvertToSecret/test_type_with_valid_fields
=== CONT  TestValidateClusterTLSConfig/test_type_with_missing_fields
=== CONT  TestValidateClusterTLSConfig/test_empty_type
--- PASS: TestBuildNamespacedName (0.00s)
    --- PASS: TestBuildNamespacedName/test_type_with_valid_fields (0.00s)
    --- PASS: TestBuildNamespacedName/test_type_with_valid_fields_and_namespaced_names (0.00s)
    --- PASS: TestBuildNamespacedName/test_type_with_non-valid_fields (0.00s)
=== PAUSE TestValidateCapiSecret/test_type_with_valid_fields
=== CONT  TestConvertToSecret/test_type_with_non-valid_fields
=== RUN   TestValidateCapiSecret/test_type_with_wrong_secret.Data[key]
--- PASS: TestValidateClusterTLSConfig (0.00s)
    --- PASS: TestValidateClusterTLSConfig/test_type_with_valid_fields (0.00s)
    --- PASS: TestValidateClusterTLSConfig/test_type_with_non-valid_field (0.00s)
    --- PASS: TestValidateClusterTLSConfig/test_type_with_missing_fields (0.00s)
    --- PASS: TestValidateClusterTLSConfig/test_empty_type (0.00s)
=== PAUSE TestValidateCapiSecret/test_type_with_wrong_secret.Data[key]
=== RUN   TestValidateCapiSecret/test_type_with_wrong_secret.Type
=== PAUSE TestValidateCapiSecret/test_type_with_wrong_secret.Type
=== CONT  TestValidateCapiSecret/test_type_with_valid_fields
=== CONT  TestExtractTakeAlongLabel/Test_with_valid_take-along-labels_label
=== CONT  TestValidateCapiSecret/test_type_with_wrong_secret.Data[key]
=== CONT  TestUnmarshal/test_type_with_wrong_secret.Data[key]
=== CONT  TestValidateCapiSecret/test_type_with_wrong_secret.Type
--- PASS: TestValidateCapiSecret (0.00s)
    --- PASS: TestValidateCapiSecret/test_type_with_valid_fields (0.00s)
    --- PASS: TestValidateCapiSecret/test_type_with_wrong_secret.Data[key] (0.00s)
    --- PASS: TestValidateCapiSecret/test_type_with_wrong_secret.Type (0.00s)
--- PASS: TestExtractTakeAlongLabel (0.00s)
    --- PASS: TestExtractTakeAlongLabel/Test_with_standard_label (0.00s)
    --- PASS: TestExtractTakeAlongLabel/Test_with_no_take-along-labels_labels (0.00s)
    --- PASS: TestExtractTakeAlongLabel/Test_with_complex_and_valid_take-along-labels_label (0.00s)
    --- PASS: TestExtractTakeAlongLabel/Test_with_valid_take-along-labels_label (0.00s)
--- PASS: TestConvertToSecret (0.00s)
    --- PASS: TestConvertToSecret/test_type_with_non-valid_fields (0.00s)
    --- PASS: TestConvertToSecret/test_type_with_valid_fields (0.00s)
--- PASS: TestUnmarshal (0.00s)
    --- PASS: TestUnmarshal/test_type_with_wrong_secret.Type (0.00s)
    --- PASS: TestUnmarshal/test_type_with_wrong_secret.Data[key] (0.00s)
    --- PASS: TestUnmarshal/test_type_with_valid_fields (0.30s)
FAIL
coverage: 28.8% of statements
FAIL    github.com/dntosas/capi2argo-cluster-operator/controllers       69.065s
FAIL
make: *** [Makefile:46: test] Error 1
```
A quick refactor and removal of the capicluster dependency seems that your logic isn't reliant on this at all from the quick look at the reconciler, just secrets.  Also, I would take advantage of structuring your tests the way ginkgo/gomega recommend to better target the need/ask of your behavior.

**Comment by siredmar** [comment]
@dntosas ready to review

**Comment by dntosas** [comment]
@siredmar started checking FYI 👀

**Comment by dntosas** [comment]
@siredmar overall LGTM

got a nit comment about vars/labels naming, I believe we can make it simpler but as you had done a lot of work already here, I suggest merging this one, probably releasing it so you can use it and then handle this stuff in a next PR :) WDYT?

**Review by dntosas — APPROVED** [review]
🔥

**Comment by siredmar** [comment]
> got a nit comment about vars/labels naming, I believe we can make it simpler but as you had done a lot of work already here, I suggest merging this one, probably releasing it so you can use it and then handle this stuff in a next PR :) WDYT?

Sure! Open a issue and we can discuss the changes if you like.

**Pull request merged**